### PR TITLE
fix Firestore C++ g3 build

### PR DESCRIPTION
**Pull request by zxu123**
Resolve conflict w.r.t. string and isnan().

This is not yet the change to make the FieldPath portable but a first step to make it build. The change w.r.t. STLPort will be in subsequent PRs.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/comparison.cc [hunk 1]
```diff
@@ -19,11 +19,10 @@
 #include <cmath>
 #include <limits>
 
-using std::isnan;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/comparison.cc]: What's the reason for this change?

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 2]
```diff
@@ -59,6 +59,23 @@ bool IsValidIdentifier(const std::string& segment) {
   return true;
 }
 
+/** A custom formatter to be used with absl::StrJoin(). */
+struct JoinEscaped {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: Have you considered using `DISABLE_GOOGLE_STRING` macro instead?

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 3]
```diff
@@ -59,6 +59,23 @@ bool IsValidIdentifier(const std::string& segment) {
...
+    const bool needs_escaping = !IsValidIdentifier(escaped);
+    if (needs_escaping) {
+      escaped.insert(escaped.begin(), '`');
+      escaped.push_back('`');
+    }
+    return escaped;
+  }
+
+  template <typename T>
+  void operator()(T* out, const std::string& segment) {
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: Does this need to be templated? i.e. would s/T* out/std::string* out/ work?

File: Firestore/core/src/firebase/firestore/util/comparison.cc (hunk 1 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/util/comparison.cc]: Possibly this: http://go/totw/119#scope-of-the-alias

Though it seems otherwise unrelated to this PR, so maybe there's a different reason?

Comment (zxu123) [reply Firestore/core/src/firebase/firestore/util/comparison.cc]: The reason is already in the description -- resolve conflict. To be precise, this error *target of using declaration conflicts with declaration already in scope*

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 2 above)
Comment (zxu123) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: It is a good ideal for building a particular app. However, may not work for library. Say a 1p app depends on Firestore and Foo, where we disable_google_string while Foo does not and both depends on Bar, where DISABLE_GOOGLE_STRING matters ... The behavior is not obviously when Bar is involved.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 3 above)
Comment (zxu123) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: This is to resolve the conflict of `std::string` and `string`. This way, both 3p and 1p build is happy. FYI: this is the suggested solution from absl team.

File: Firestore/core/src/firebase/firestore/util/comparison.cc (hunk 1 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/comparison.cc]: Yes, I'm not sure where the unqualified `isnan` is coming from, does something include `<math.h>`? Likely not worth digging into, though.
